src/hopen/syntax/mustache.cljc

```diff
@@ -0,0 +1,434 @@
+                               (conj-last (pop stack) $)) ;; move last context into parent context
+
+                             ;; Add the tag to the active context
+                             :else
+                             (conj-last stack tag))))
+                       [[]]  ;; start with stack with a single empty top level context
+                       tags)]
+
+    ;; We've processed all the tags.
+    ;; If all went well, we should have exactly one open contet that represents a valid nested sections.
+    ;; If the user left some sections unclosed, we'd end up with multiple contexts here.
+    ;; We'll place these into the root context for now.
+    ;; This error will be detected and reported in `ast-sanity-check`.
+    (if (< 1 (count result))
+      (apply conj (first result) (rest result))
+      (first result))))
+
+(defn ast-sanity-check [ast]
+  (doseq [node (tree-seq vector? seq ast)]
+    (when (map? (first node))
+      (let [open-tag (first node)
+            close-tag (last node)]
+        (if (not= (:section-name open-tag) (:section-name close-tag))
+          (throw (ex-info (str "Section tag \"" (get-in open-tag [:groups 0])
+                               "\" opened on line " (:lineno open-tag)
+                               " is never closed") {}))))))
+  ast)
+
+(defn ast-node->hopen-node
+  [node]
+
+  (cond
+    (string? node) node
+    (map? node) (cond
+                  (= :section-open (:type node)) node
+                  (= :var-ref (:type node))
+                  (if (:skip-escape node)
+                    (backtick/template
+                     (ctx-lookup ~(keyword (:var-name node))))
+                    (backtick/template
+                     (escape-html (ctx-lookup ~(keyword (:var-name node))))))
+                  (= :section-close (:type node)) node
+                  (= :inverted-open (:type node)) node
+                  (= :comment (:type node)) nil
+                  (= :partial (:type node)) nil
+                  (= :delim-change (:type node)) nil
+                  :else (throw (ex-info (str "Does not how how to deal with tags of type: " (:type node)) {})))
+
+    (vector? node)
+    (cond
+      (= :section-open (:type (first node))) (let [contents (->> node
+                                                                 (drop 1)
+                                                                 (drop-last 1)
+                                                                 (into []))]
+                                               (backtick/template
+                                                (b/let [data (ctx-lookup ~(keyword (:section-name (first node))))]
+                                                  [(b/if (and data
+                                                              (not (empty? data)))
+                                                     [(b/if (or (list? data)
+                                                                (vector? data))
```

Comment:
I think that what you really need is `(seq` and `(not (seq`.

Reply:
I'm not sure where you're suggesting for `(seq` and `(not (seq` to go. I think this code is behaving the way it's supposed to.

What this is trying to do is:
1) if `data` is not present (nil) or an empty collection, skip processing contents
2) if `data` is a map (single data item) , it needs to be processed one way
3) If `data` is it's a list/vector (multiple data items), it should be processed another way.

Reply:
Thanks for the explanation, I was not sure about what was intended and wanted to ask.

`(and data (not (empty? data)))` is idiomatically written `(seq data)` in Clojure.

In your second test, a set `#{:a :b}` will be treated in the `else`, that might be a problem.
You might use `map?` instead.

Could you explain that part?
> if data is a map (single data item)

Reply:
From the mustache spec:
> **Non-False Values**
> 
> When the value is non-false but not a list, it will be used as the context for a single rendering of the block.
> 
> Template:
> 
> {{#person?}}
>   Hi {{name}}!
> {{/person?}}
> Hash:
> 
> {
>   "person?": { "name": "Jon" }
> }
> Output:
> 
> Hi Jon!

I guess since the variable reference syntax doesn't have any notion of containment, this is the way it uses to basically do `(get-in ctx [:person? :name])`.

Reply:
Yes, it's idiomatic to use `(seq data)` instead of `(not (empty? data))` or `(not-empty data)` to ask if a collection is empty. However, semantically, `seq` means to try to get a sequence representation of something. It's interesting that you can also use it to answer if a collection is empty or not, but using it that way is mixing semantics and does not contribute to readability.

Though I prefer `empty?` and `not-empty` which says exactly what it does.

Reply:
I forgot that the value could be a boolean. Your approach was correct.